src/player.js

```diff
@@ -1,219 +1,180 @@
-   * @param {HTMLMediaElement} mediaElement The HTML <code>&lt;audio&gt;</code>
-   *   or <code>&lt;video&gt;</code> element to associate with the
-   *   {@link Peaks} instance.
+   * @param {Adapter} adapter The player adapter.
    */
 
-  function Player(peaks, mediaElement) {
+  function Player(peaks, adapter) {
     var self = this;
 
     self._peaks = peaks;
-    self._listeners = [];
-    self._mediaElement = mediaElement;
-    self._duration = self.getDuration();
-    self._isPlaying = false;
 
-    self._addMediaListener('timeupdate', function() {
-      self._peaks.emit('player_time_update', self.getCurrentTime());
-    });
+    validateAdapter(adapter);
+    self._adapter = adapter;
 
-    self._addMediaListener('play', function() {
-      self._isPlaying = true;
-      self._peaks.emit('player_play', self.getCurrentTime());
-    });
-
-    self._addMediaListener('pause', function() {
-      self._isPlaying = false;
-      self._peaks.emit('player_pause', self.getCurrentTime());
-    });
-
-    self._addMediaListener('seeked', function() {
-      self._peaks.emit('player_seek', self.getCurrentTime());
-    });
+    self._adapter.init(self);
 
-    self._addMediaListener('canplay', function() {
-      self._peaks.emit('player_canplay', self);
-    });
-
-    self._addMediaListener('error', function(event) {
-      self._peaks.emit('player_error', event.target.error);
-    });
-
-    self._interval = null;
-  }
-
-  /**
-   * Adds an event listener to the media element.
-   *
-   * @private
-   * @param {String} type The event type to listen for.
-   * @param {Function} callback An event handler function.
+    /**
+   * Cleans up the player object.
    */
 
-  Player.prototype._addMediaListener = function(type, callback) {
-    this._listeners.push({ type: type, callback: callback });
-    this._mediaElement.addEventListener(type, callback);
-  };
-
-  /**
-   * Cleans up the player object, removing all event listeners from the
-   * associated media element.
-   */
-
-  Player.prototype.destroy = function() {
-    for (var i = 0; i < this._listeners.length; i++) {
-      var listener = this._listeners[i];
-
-      this._mediaElement.removeEventListener(
-        listener.type,
-        listener.callback
-      );
-    }
+    self.destroy = function() {
+      self._adapter.destroy();
+    };
 
-    this._listeners.length = 0;
-
-    if (this._interval !== null) {
-      clearTimeout(this._interval);
-      this._interval = null;
-    }
-
-    this._mediaElement = null;
-  };
-
-  Player.prototype.setSource = function(source) {
-    this._mediaElement.setAttribute('src', source);
-  };
-
-  Player.prototype.getSource = function() {
-    return this._mediaElement.src;
-  };
-
-  Player.prototype.getCurrentSource = function() {
-    return this._mediaElement.currentSrc;
-  };
-
-  /**
+    /**
    * Starts playback.
    */
 
-  Player.prototype.play = function() {
-    this._mediaElement.play();
-  };
+    self.play = function() {
```

Comment:
I'd prefer to continue to use the `Player.prototype.play = function()` style, here and in MediaElementPlayer, if we can.

Reply:
Done, switched to prototype style for Player and MediaElementPlayer.
As a consequence the adapter validation must be done on the adapter object and its prototype. Added a `getAllPropertiesFrom(adapter)` to this purpose.